crates/core_simd/src/swizzle.rs

```diff
@@ -364,4 +364,131 @@ where
+    /// from the input vectors. You must specify it explicitly.
+    ///
+    /// ```
+    /// # #![feature(portable_simd)]
+    /// # use core::simd::Simd;
+    /// let x = Simd::from_array([0, 1, 2, 3]);
+    /// let y = Simd::from_array([4, 5, 6, 7]);
+    /// let z = x.concat_to::<8>(y);
+    /// assert_eq!(z.to_array(), [0, 1, 2, 3, 4, 5, 6, 7]);
+    /// ```
+    ///
+    /// Will be rejected at compile time if `LANES * 2 != DOUBLE_LANES`.
+    #[inline]
+    #[must_use = "method returns a new vector and does not mutate the original inputs"]
+    pub fn concat_to<const DOUBLE_LANES: usize>(self, other: Self) -> Simd<T, DOUBLE_LANES>
+    where
+        LaneCount<DOUBLE_LANES>: SupportedLaneCount,
+    {
+        const fn concat_index<const DOUBLE_LANES: usize>(lanes: usize) -> [Which; DOUBLE_LANES] {
+            assert!(lanes * 2 == DOUBLE_LANES);
+            let mut index = [Which::First(0); DOUBLE_LANES];
+            let mut i = 0;
+            while i < lanes {
+                index[i] = Which::First(i);
+                index[i + lanes] = Which::Second(i);
+                i += 1;
+            }
+            index
+        }
+        struct Concat;
+        impl<const LANES: usize, const DOUBLE_LANES: usize> Swizzle2<LANES, DOUBLE_LANES> for Concat {
+            const INDEX: [Which; DOUBLE_LANES] = concat_index::<DOUBLE_LANES>(LANES);
+        }
+        Concat::swizzle2(self, other)
+    }
+
+    /// For each lane `i`, swaps it with lane `i ^ SWAP_MASK`.
+    ///
+    /// Also known as `grev` in the RISC-V Bitmanip specification, this is a powerful
+    /// swizzle operation that can implement many common patterns as special cases.
+    ///
+    /// ```
+    /// # #![feature(portable_simd)]
+    /// # use core::simd::Simd;
+    /// let x = Simd::from_array([0, 1, 2, 3, 4, 5, 6, 7]);
+    /// // Swap adjacent lanes:
+    /// assert_eq!(x.general_reverse::<1>().to_array(), [1, 0, 3, 2, 5, 4, 7, 6]);
+    /// // Swap lanes separated by distance 2:
+    /// assert_eq!(x.general_reverse::<2>().to_array(), [2, 3, 0, 1, 6, 7, 4, 5]);
+    /// // Swap lanes separated by distance 4:
+    /// assert_eq!(x.general_reverse::<4>().to_array(), [4, 5, 6, 7, 0, 1, 2, 3]);
+    /// // Reverse lanes, within each 4-lane group:
+    /// assert_eq!(x.general_reverse::<3>().to_array(), [3, 2, 1, 0, 7, 6, 5, 4]);
+    /// ```
+    ///
+    /// Commonly useful for horizontal reductions, for example:
+    ///
+    /// ```
+    /// # #![feature(portable_simd)]
+    /// # use core::simd::Simd;
+    /// let x = Simd::from_array([0u32, 1, 2, 3, 4, 5, 6, 7]);
+    /// let x = x + x.general_reverse::<1>();
+    /// let x = x + x.general_reverse::<2>();
+    /// let x = x + x.general_reverse::<4>();
+    /// assert_eq!(x.to_array(), [28, 28, 28, 28, 28, 28, 28, 28]);
+    /// ```
+    #[inline]
+    #[must_use = "method returns a new vector and does not mutate the original inputs"]
+    pub fn general_reverse<const SWAP_MASK: usize>(self) -> Self {
```

Comment:
imho this should include something like `lanewise` in it's name, since I expect Rust to gain a bitwise `grev` at some point and we'd want the bitwise simd and scalar integer operations to have matching names (probably `gen_rev` or `generalized_reverse` or `grev`?).

Reply:
For something that effectively implements a very specific xor-striding pattern, `general_reverse` is a non-descript name.

Reply:
well, `grev` *is* what that particular bitwise op has become named, thanks to RISC-V afaict.

Reply:
grev -- general bit reverse

Reply:
Just because an ISA has picked a terrible name doesn't mean we need to copy it.

Reply:
well, it gets the name because it's a generalization of bitwise reverse and all byte swaps for reversing endian -- e.g. `grev(v, 0x3F)` is 64-bit bitwise reverse, `grev(v, 0x38)` reverses all bytes in a 64-bit value, `grev(v, 0x18)` reverses bytes in each 32-bit part, `grev(v, 0x8)` reverses bytes in each 16-bit part, etc.

imho `grev` is the name people use even if it wasn't the best originally, so we should use it for consistency and searchability.

Reply:
`swizzle_to_xor_indexes`?

Reply:
`reverse_pow2_lane_groups`?  Conceptually this operation performs the reversal of blocks of `k` lanes within `n`-lane groups, where `k` and `n` are both powers of two, `k ≤ n ≤ LANES`, and the choice of `k` and `n` is determined uniquely up to operation uniqueness by choosing where index `0` will be swizzled to (it’ll be exactly `SWAP_MASK`).

Reply:
> `reverse_pow2_lane_groups`?

afaict `grev` is actually more powerful than that, it can do any arbitrary combination of those `k`-`n` reversals for arbitrary `k` and `n`.

e.g. `grev(v, 5)` is equivalent to `simd_swizzle!(v, [5, 4, 7, 6, 1, 0, 3, 2])` which is a combination of reversing adjacent pairs (blocks of 1) and swapping blocks of 4.